daffodil-core/src/main/scala/org/apache/daffodil/runtime2/Runtime2DataProcessor.scala

```diff
@@ -0,0 +1,211 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.runtime2
+
+import java.io.File
+import java.io.InputStream
+import java.io.OutputStream
+
+import org.apache.daffodil.api.DFDL
+import org.apache.daffodil.api.DaffodilTunables
+import org.apache.daffodil.api.DataLocation
+import org.apache.daffodil.api.ValidationMode
+import org.apache.daffodil.externalvars.Binding
+import org.apache.daffodil.processors.Failure
+import org.apache.daffodil.processors.ProcessorResult
+import org.apache.daffodil.processors.Success
+import org.apache.daffodil.processors.VariableMap
+import org.apache.daffodil.processors.WithDiagnosticsImpl
+import org.apache.daffodil.processors.parsers.ParseError
+import org.apache.daffodil.processors.unparsers.UnparseError
+import org.apache.daffodil.util.Maybe
+import org.apache.daffodil.util.Maybe.Nope
+import os.Pipe
+
+/**
+ * Effectively a scala proxy object that does its work via the underlying C-code.
+ * Will need to consider how to use features of underlying C-code to get infoset,
+ * walk infoset, generate XML for use by TDML tests.
+ */
+class Runtime2DataProcessor(executableFile: os.Path) extends DFDL.DataProcessorBase {
+  /**
+   * Returns a data processor with all the same state, but the validation mode changed to that of the argument.
+   *
+   * Note that the default validation mode is "off", that is, no validation is performed.
+   */
+  override def withValidationMode(mode: ValidationMode.Type): DFDL.DataProcessor = ???
+
+  override def withTunable(name: String, value: String): DFDL.DataProcessor = ???
+
+  override def withTunables(tunables: Map[String, String]): DFDL.DataProcessor = ???
+
+  override def withExternalVariables(extVars: Map[String, String]): DFDL.DataProcessor = ???
+
+  override def withExternalVariables(extVars: File): DFDL.DataProcessor = ???
+
+  override def withExternalVariables(extVars: Seq[Binding]): DFDL.DataProcessor = ???
+
+  override def validationMode: ValidationMode.Type = ???
+
+  override def getTunables(): DaffodilTunables = ???
+
+  override def save(output: DFDL.Output): Unit = ???
+
+  override def variableMap: VariableMap = ???
+
+  override def setValidationMode(mode: ValidationMode.Type): Unit = ???
+
+  override def setExternalVariables(extVars: Map[String, String]): Unit = ???
+
+  override def setExternalVariables(extVars: File): Unit = ???
+
+  override def setExternalVariables(extVars: File, tunable: DaffodilTunables): Unit = ???
+
+  override def setExternalVariables(extVars: Seq[Binding]): Unit = ???
+
+  override def setTunable(tunable: String, value: String): Unit = ???
+
+  override def setTunables(tunables: Map[String, String]): Unit = ???
+
+  /**
+   * Returns an object which contains the result, and/or diagnostic information.
+   */
+  def parse(input: InputStream): ParseResult = {
```

Comment:
This feels odd to me. I would expect the runtime2 to not actually have a parse method. It would just generate an executable/library and users would have to know how to use that. So the API sort of changes depending on what backends are made avialable. We may need to rethink how things like the API and CLI work. For example, if this runtime2 backend were on the class path, you might expect something like this to be available:

daffodil runtime2 -s schema.dfdl.xsd -o schema.bin

This would generate a schema.bin executable that could then be used to do the actual parse unparse. So the concept of a DataProcessor in runtime2 migh not exist. And it would be purely a runtime1 concept?

The fact that so many of these methods are unimplemented says to be me need to put some thought into how these backends will work in practice.

Reply:
Agree we need  to think this through.

Some requirements: 
* use an alternate back-end, but still run TDML tests against it interactively. (C-code generation, C-compiler, and invoking of it all transparent to the user.)

That requirement is what motivates the runtime2 library having a walker that outputs XML and inputs XML, and a TDML processor that glues it all together. This motivates the creation of things like DataProcessor even though the concept makes little sense for runtime2, it allows reuse of more code in creating the test interface for TDML. 

If you have that requirement, the same mechanisms can be used to provide the CLI with the same "snap compiling" capability so that you can have daffodil parse and daffodil unparse for the CLI that "work-alike" across the back-ends.  

A separate requirement is to be able to generate a ".a" or ".so" C-callable library and matching ".h" include files that implements parse and unparse C APIs. The generated code would be one library distinct from the static code runtime2 library (which is likely to be quite large eventually).  Creating this generated library feels like what the CLI should do when using runtime 2 backend and invoking the save parser option.

Reply:
Regarding the TDML requirement, I think this can all be handled via the TDMLDFDLProcessor API. Perhaps different backends will also want different TDMLDFDLlProcessor implementations. The runtime2 implementation would generate and compile code, execute the binary with the appropriate command line arguments based on the test configuration, capture the output, and return it to the TDML Runner to check for correctness. We already have support for tests to define which implementation they work for and the ability to switch different implementations. So it's just a matter of defining this new implementation. So there is certainly a concept of a DFDLProcessor specific to runtime2, but it doesn't necessarily need to match the DataProcessor API that runtime1 uses. It really just needs to match the TDMLDFDLProcessor API.

Agreed on the second requirment. Generating a shared object with some standard API that users can easily call ultimately seems like best approach. Generating another CLI implemented in C is easy for testing, but utlimately likely not how people will want to use it in the real world. I could imagine an implementation of the TDMLDFDLProcessor could use JNI to dynamically load these generated libraries and hook into the right function calls to parse/unparse.

Regarding the save-parser, I'm wonder if options really become part of the runtime. Different runtimes can determine which options they expect and what actions to take.

For example, the --debug/--trace options might not make sense with runtime2. Debugging the generated C code likely just means someone should compile the code and execute it with GDB, for example. And the save-parser command for the runtime2 likely wants additional parameters specific to it, like c compiler options, or output diretory for header files. The runtimes could very well require very different parameters. I'm not sure the following is a good idea, but maybe the current CLI wants to become something like this:

```
usage: daffodil <runtime> <runtime subcommands/options>

daffodil runtime1 parse --schema foo.dfdl.xsd ...
daffodil runtime2 --schema foo.dfdl.sd -copts=-Wfoo --incldir include/
```
Note that different runtimes don't necessarily need to have subcommands. Runtime1 would likely have many of the existing subcommands.

Anoher option would be to keep the root level parse/unparse/save-parser options, and just have a --runtime option, and depending on which runtime option you select different options will become available. That might be a bit easier and more familiar, but the usage likely gets more complicated since availability of parameters changes depending on the runtime. I'm not sure how well scallop can handle something like that.

 
Another option to the save-parser thing
To be even more pluggable, it would also be interesting if the Daffodil Scala CLI could ask the different backends for their CLI options.

Reply:
Arguably, the current CLI is a daffodil-runtime1 CLI, and a new runtime should create its own tools. If they can share library code great, but perhaps we don't even hook runtime2 into the existing CLI at all?

Reply:
A separate subcommand such as generate doesn'ts seem unreasonable. Though, I can imagine that if we ended up with another runtime that generated something else (eg. python, java, vhdl) that the generate command would take very different options, which could cause confusion. Have a separate command per runtime solves this, but I think maybe makes usability a bit harder, especially since the backend names aren't very obvious (runtime1) when compared to "parse/unparse".

Maybe the best option is to just make our CLI runtime1 focused as Mike suggestes, but then we still need a way to execute runtime2. Maybe it's a separate CLI tool with a completely different name, and all new runtimes will get a unique CLI. This too might be worth a larger discussion on the dev list. How are people expected to use and trigger different backends